### PR TITLE
add authorization

### PR DESCRIPTION
**Pull request by agnbart**

### CODE REVIEW

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
We need to restrict unauthorized requests

### Requirements
1. It should check if incoming request contains authorization header called `Authorization`.
2. Incoming request `Authorization` header value should be equal `pss-this-is-my-secret`
3. Throw `Unauthorized` response status when conditions are not met

### Review

#### index.ts
The middleware `checkAuthorization` is correctly added to the express app using `app.use(checkAuthorization);`. This ensures that the authorization check will be performed for all incoming requests.

#### utils/check-authorization.ts
The `checkAuthorization` function correctly checks if the incoming request contains an `Authorization` header and if its value is equal to `pss-this-is-my-secret`.

However, there are a couple of issues with the current implementation:

1. The response status when the conditions are not met is 400 (Bad Request), but according to the requirements, it should be `Unauthorized` which corresponds to status code 401. Therefore, the lines where `res.status(400).send('Invalid authorization');` and `res.status(400).send('Authorization header missing');` are used should be changed to `res.status(401).send('Invalid authorization');` and `res.status(401).send('Authorization header missing');` respectively.

2. The middleware does not call the `next()` function when the authorization check passes. This means that the request handling will stop at this middleware and no subsequent middleware or route handlers will be executed. To fix this, add a call to `next()` in the branch of the if statement where the authorization check passes.

3. The middleware sends a response with the message 'Authorization header received' when the authorization check passes. This is not required by the task and might interfere with the responses of other route handlers. It would be better to remove this line.

Here's how the corrected code might look:

```diff
export const checkAuthorization = (req: Request, res: Response, next: NextFunction) => {
    const authorizationHeader = req.rawHeaders.find((element, index) => {
        return element.toLowerCase() === 'authorization' && req.rawHeaders[index + 1];
    });

    if (authorizationHeader) {
        const authorizationValue = req.rawHeaders[req.rawHeaders.indexOf(authorizationHeader) + 1];
        if (authorizationValue && authorizationValue === "pss-this-is-my-secret") {
-            console.log('Authorization header:', authorizationValue);
-            res.send('Authorization header received');
+            next();
        } else {
-            res.status(400).send('Invalid authorization');
+            res.status(401).send('Invalid authorization');
        }
    } else {
-        res.status(400).send('Authorization header missing');
+        res.status(401).send('Authorization header missing');
    }
}
```

In conclusion, the code changes partially cover the task requirements. The main functionality of checking the `Authorization` header is implemented correctly, but the response status code does not match the requirements and the middleware does not correctly pass control to the next middleware or route handler.

**Review by NerdbordMentor — APPROVED** [review]
## Code Review Report

### Task goal
The goal of the task was to restrict unauthorized requests.

### Requirements
1. It should check if the incoming request contains an authorization header called `Authorization`.
2. The incoming request `Authorization` header value should be equal to `"pss-this-is-my-secret"`.
3. Throw `Unauthorized` response status when conditions are not met.

### Code changes review

#### index.ts
The middleware `checkAuthorization` was added to the express app, which is a correct step towards achieving the task goal.

#### utils/check-authorization.ts
The `checkAuthorization` function was created to handle the authorization of incoming requests. It checks if the incoming request contains an authorization header and if the value of the header is equal to `"pss-this-is-my-secret"`. If these conditions are not met, it sends a response with a status of 401 (Unauthorized), which is in line with the task requirements.

### Conclusion
The code changes cover the task requirements. The `checkAuthorization` middleware function correctly checks for the presence and value of the `Authorization` header in incoming requests and sends an Unauthorized response when these conditions are not met.
